### ISSUE TITLE
Improving Small Object Detection in Large Images Using Tiling

### ISSUE THREAD

**Issue by Uddeshya1052** [ultralytics/yolov5#13520]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Description
I am working on detecting very small objects (e.g., screws) within large images (~4000x3000 pixels) of a perforated tank. However, I am facing challenges with low detection accuracy, likely due to the small object size relative to the full image.

I believe tiling could be a good solution, but I am unsure of the best approach to integrate it into the training pipeline. Specifically:

Should I label the full images first and then apply tiling, or should labeling be done after tiling?
What is the recommended way to ensure consistency between tiles and the original annotations?
Are there any best practices or existing implementations that could help with tiling for small object detection?
Any guidance, references, or example scripts would be greatly appreciated.

Context
Image Size: ~4000x3000
Objects: Very small (screws, etc.)
Model: YOLO-based detection (but open to suggestions)
Thank you in advance for your help! 🚀

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Uddeshya1052, thank you for your interest in YOLOv5 🚀! It sounds like you're tackling an interesting challenge in small object detection within large images. Please explore the following resources to potentially help address your question:

### Resources

1. ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) - Start from our Tutorials to better understand YOLOv5 capabilities, including tips for preparing datasets and optimizing detection.
2. [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) - Learn how to improve model performance through data augmentation and anchoring techniques.
3. [Discussions](https://github.com/ultralytics/yolov5/discussions) - Join the community to exchange tips about small object detection or tiling methods.

### Recommendations
If tiling is your planned approach:
1. Annotating **full images first** is generally recommended to ensure all objects are accounted for before splitting into tiles.
2. Ensure tiling processes maintain annotation consistency—each tile must contain properly mapped labels relative to the objects within it. Open-source tools like `roboflow` or custom scripts can automate this step.
3. Consider **data augmentation**, ensuring that the tiles resemble diverse real-world conditions and preventing overfitting to specific sizes or resolutions.

As YOLOv5 has high flexibility, you might also experiment with model hyperparameters like anchor sizes or train-scaled variants (e.g., YOLOv5x vs YOLOv5s) for detecting smaller objects 🪛. Check out [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/) for guidance.

### Additional Resources
YOLOv5 may be used in combination with diverse environments where tiling workflows can be implemented:
- **Notebooks**: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- For automation, consider a **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

### Requirements
Ensure you are running:
- [Python>=3.8.0](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed
- [PyTorch>=1.8](https://pytorch.org/get-started/locally/)
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

### Next Steps
If this is a ❓ Question and you'd like further help, please:
- Share additional details about your current progress, tiling script (if used), or results so far.
- Provide examples of the images and labeling format to better assist.

If this is a 🐛 Bug Report regarding YOLOv5 behavior, include a **minimum reproducible example (MRE)** for debugging.

This is an automated response to assist you efficiently 🤖. An Ultralytics engineer will review and provide additional guidance soon 🔍. Thank you again for your enthusiasm and for using YOLOv5! 🚀

**Comment by pderrenger** [comment]
@Uddeshya1052 for small object detection in large images with YOLOv5, we recommend:

1. **Label Full Images First**: Annotate the original high-resolution images, then use a tiling script to split both images and labels into smaller tiles (e.g., 640x640). This maintains annotation consistency automatically.

2. **Tiling Implementation**:
```python
from utils.dataloaders import LoadImagesAndLabels

dataset = LoadImagesAndLabels(path, imgsz=640, augment=True)
# Enable "rectangular training" with mosaic augmentation in your train.py call:
# python train.py --rect --img 640
```

3. **Key Adjustments**:
- Use `--img 640` (or larger)
- Reduce anchor sizes in your YAML
- Increase `--iou-thres` slightly

For implementation details, see the Ultralytics YOLOv5 documentation on training with mosaic augmentation and rectangular training. This approach effectively handles small objects while maintaining original annotation relationships.

**Comment by Uddeshya1052** [comment]
@pderrenger Thank you for your answer! How can I get the tiling script? Is there an official source or example code available?

**Comment by pderrenger** [comment]
@Uddeshya1052 for tiling in YOLOv5, we recommend using the built-in `LoadImagesAndLabels` dataset class with `rectangular training` (via `--rect` flag) and mosaic augmentation. This automatically handles image/label tiling during training without needing separate scripts. See the [LoadImagesAndLabels implementation](https://github.com/ultralytics/yolov5/blob/master/utils/dataloaders.py) and [training documentation](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) for details. For custom tiling needs, you can modify this class while maintaining label alignment.
